### ISSUE TITLE
Improve the panning and smoothness of panoramas with large resolutions

### ISSUE THREAD

**Issue by vinaiah** [mpetroff/pannellum#911]
Hi - we have been using this amazing piece of software for a while now and generally had no issues with it. Lately we have started supporting very large panoramas - upto 32K sometimes 

Foe eg. : These tours 
- https://teliportme.com/virtualtour/7ca43727 
- https://teliportme.com/virtualtour/652e8453
- https://teliportme.com/virtualtour/4749da87

We have noticed that our panning gets sluggish and is not as smooth as for example viewers like 

- https://roundme.com/tour/444946/view/1531158/
- https://kuula.co/post/7T7ld

While our customers love that we have this focus on the right kind of projection/less distortion and our ability to provide high quality resolution; they have complained before about the sluggishness of the viewer. We have followed all the best practices but are unable to reproduce a similar smooth scroll/panning in our viewer. 

Your help in this would be great. 

Also how can we support your efforts ?

**Comment by mpetroff** [comment]
TL;DR: Jank with multiresolution images is caused by garbage collection. This could be improved by reducing heap usage, but I have a more complete rewrite planned.

This is a complicated subject, although I'll note that there are only performance issues / jank when Pannellum's multiresolution format is used (although this is obviously necessary for quick load times and large panoramas).

I'll start with some historical context. Pannellum's multiresolution support dates back to 2013, a time when JavaScript was still essentially always single-threaded and synchronous. WebGL is obviously even older. Thus, the WebGL API can only be accessed from the main thread and involves synchronous calls. Furthermore, Flash was still widely used for panoramas when I added Pannellum's multiresolution support, so I tried to make the multiresolution support backwards-compatible with the tiling formats used by some of the commonly used Flash-based viewers.

In my experience, the jank sometimes seen in Pannellum is due to the browser garbage collecting some of the objects used to load and display multiresolution tiles (which is why it's not an issue for equirectangular images). I've made attempts over the years to improve this, but I've only been partially successful. For example, only one tile is ever loaded per frame in an attempt to avoid dropping frames (faster machines support loading multiple tiles per frame, but this isn't done due to difficulties arising from the synchronous, single-threaded operations). Both Roundme and Kuula also drop frames due to garbage collection, just not as often, since they don't use as much heap memory as Pannellum does. Reducing Pannellum's heap usage would reduce the need for garbage collection, which would reduce the occurrence of dropped frames; however, due to the rewrite I have planned, I'm not sure I want to invest too much time into this (@vinaiah contact me in private if you want to discuss sponsored development).

My semi-long-term plan is to completely rewrite Pannellum's multiresolution support, probably sometime next year. The goals are to improve performance, reduce bandwidth usage, and simplify file management for serving multiresolution panoramas. However, this plan depends on both WebGPU and JPEG XL, neither of which have more than experimental browser support at the moment.

WebGPU has been designed with asynchronous, multi-threaded code in mind, so it should alleviate some of the issues caused by WebGL's synchronous, single-threaded execution model. Furthermore, I'm considering using Web Assembly for the implementation, since this allows for explicit control of memory and thus can completely eliminate the need for garbage collection. However, the use of Web Assembly may not be necessary for this, since moving the tile loading off of the main rendering thread may be sufficient, since garbage collection would then not block rendering. This portion of the rewrite would lead to improved performance and ideally make dropped frames, and thus jank, an extremely rare occurrence.

The bandwidth reduction and file management simplification would result from the use of JPEG XL. One of JPEG XL's goals is to provide responsive image support in a single image, to eliminate the need for storing multiple copies of an image at different resolutions for use with responsive websites. It accomplishes this by internally creating an image pyramid, similar to Pannellum's current multiresolution format; however, it is more efficient at this, since its higher resolution tiles depend on information in the lower resolution tiles instead of being completely independent. This, along with JPEG XL's coding efficiency improvements, would reduce bandwidth usage (and server storage usage). By using HTTP range requests, the tiles should be able to be loaded directly from the single JPEG XL file, which simplifies file management, since individual tiles no longer need to be stored on disk.

The final aspect to reduce bandwidth usage is to use a more efficient image projection. Currently, Pannellum uses a Gnomonic cube map (as do many other viewers), where each cube face is then tiled. As this is not an equal-area projection, the pixel density is different in different parts of the image, which is wasteful. The ideal projection would be equal-area (or approximately equal-area), to eliminate wasted pixels; have low angular distortion and no cusps, to make lossy image compression more efficient; and have low computational complexity, for rendering performance. I've done some [recent work on a square equal-area map projection](https://arxiv.org/abs/2008.13670) to this end, although I still need to do a more careful evaluation of these square projections and various [efficient cube map projections](https://doi.org/10.1007/s00371-019-01708-4) before deciding on a particular format.

Since I'm discussing future plans, the idea is that the new renderer backend, which would also support equirectangular images, would be added in a future 2.x release. This release would still include the old backend for browsers without WebGPU / JPEG XL support and for supporting the old multiresolution format, but the old multiresolution format and cubemap support would be deprecated. For Pannellum 3.0, the front-end portion of the viewer would be updated to use ES6, and the deprecated formats would be dropped. This should go without saying, but these future plans, and the rough timeline I mentioned, are subject to change.

**Comment by vinaiah** [comment]
Thanks for the detailed answer. 

I am glad to know that you are on top of this; I will contact you for a private implementation via email (  have also sent you a linkedin request in case my email ends up in junk)

**Comment by mpetroff** [comment]
Significant performance improvements for multiresolution panoramas were made in 5eab5366f6987a4abdf648cbf5f7a243f14faf28 and 3a10b89937bc993571b0aa58e34ba44d4c9584fc. Thank you to [TeliportMe](https://teliportme.com/) for sponsoring these improvements.

**Comment by jonanderdev** [comment]
Hello @mpetroff 

First thank you for this amazing tool!
Quick doubt, is there a plan for the next release? 
I see there are some improvements for multires like this previous commit: https://github.com/mpetroff/pannellum/commit/5eab5366f6987a4abdf648cbf5f7a243f14faf28 

Just wondering as we're currently using the npm 2.5.6 and wonder when/if was gonna be updated with that latest changes.

Thanks in advance!

**Comment by mpetroff** [comment]
@jonanderdev  When it's ready and I have time for proper release testing and quality assurance; no one's stopping you from using the latest Git version.
